src/Core/src/Hosting/ImageSources/ImageSourceToImageSourceServiceTypeMapping.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+using TypePair = (System.Type ImageSource, System.Type ImageSourceService);
+
+namespace Microsoft.Maui.Hosting
+{
+	internal sealed class ImageSourceToImageSourceServiceTypeMapping
+	{
+		private static readonly ConcurrentDictionary<IImageSourceServiceCollection, ImageSourceToImageSourceServiceTypeMapping> s_instances = new();
+
+		internal static ImageSourceToImageSourceServiceTypeMapping GetInstance(IImageSourceServiceCollection collection) =>
+			s_instances.GetOrAdd(collection, static _ => new ImageSourceToImageSourceServiceTypeMapping());
+
+		private ConcurrentDictionary<Type, Type> _typeMappings { get; } = new();
+
+		public void Add<TImageSource>() where TImageSource : IImageSource =>
+			_typeMappings[typeof(TImageSource)] = typeof(IImageSourceService<TImageSource>);
+
+		public Type FindImageSourceType(Type imageSourceType) =>
+			FindImageSourceToImageSourceServiceTypeMapping(imageSourceType).ImageSource;
+
+		public Type FindImageSourceServiceType(Type imageSourceType) =>
+			FindImageSourceToImageSourceServiceTypeMapping(imageSourceType).ImageSourceService;
+
+		private TypePair FindImageSourceToImageSourceServiceTypeMapping(Type type)
+		{
+			Debug.Assert(typeof(IImageSource).IsAssignableFrom(type));
+
+			// If there's an exact match for the type, just return it.
+			if (_typeMappings.TryGetValue(type, out var imageSourceService))
+			{
+				return (type, imageSourceService);
+			}
+
+			List<TypePair> matches = new();
+			foreach (var mapping in _typeMappings)
+			{
+				var imageSource = mapping.Key;
+				if (imageSource.IsAssignableFrom(type) || type.IsAssignableFrom(imageSource))
```

Comment:
This allows for matches to more derived types right? Doesn't the original lookup only iterate through less derived types?

I'm imagining that because this lookup system depends on the order of `matches`, that `InvalidOperationException($"Ambiguous` can be hit if the frontmost matches are both more derived types of `type`.
e.g.
`MyCustomImageSource : IFileImageSource`
`MyLargeCustomImageSource : MyCustomImageSource`
`MySmallCustomImageSource : MyCustomImageSource`
Services added for `MyLargeCustomImageSource`, `MySmallCustomImageSource`, `IFileImageSource`
and user looks for the service associated with `MyCustomImageSource`
If `matches[0]` and `matches[1]` are `MyLargeCustomImageSource` and `MySmallCustomImageSource`, the exception would be hit.

But maybe this scenario is too contrived?

Reply:
> This allows for matches to more derived types right? Doesn't the original lookup only iterate through less derived types?

You're right, there are differences between my PR and the original implementation.

In the scenario you descrived:
- When `AddService< MyLargeCustomImageSource >(...)` is called, it will register a singleton with `IImageSourceService<MyLargeCustomImageSource>` service descriptor.
- The `ImageSourceServiceProvider.GetImageSourceType` method will return `IFileImageSource` for all those classes in the original implementation. The code in this PR would return the input value back for `MyLargeCustomImageSource`, `MySmallCustomImageSource`, and `IFileImageSource` (the services are registered with exactly these types) and it would throw for `MyCustomImageSource` because there isn't any service registered for `MyCustomImageSource` and there are multiple registered image sources that are related to it.
- the `ImageSourceServiceProvider.GetImageSourceServiceType` will first check if there is a service descriptor for the input type (`IImageSourceService<MyLargeCustomImageSource>` and so on) and it will return this descriptor if it is there. This would be the case for `MyLargeCustomImageSource`, `MySmallCustomImageSource`, and `IFileImageSource` and these types would work just fine. If the descriptor isn't available, the original code would default to `IImageSourceService<IFileImageSource>` and try to locate this service. My PR would throw instead of this fallback.

I will turn your scenario into a unit test and I'll fix my PR.

Reply:
I updated the code to match the previous implementation much more closely. In the end, it basically "just" moves all the logic to the `Add` method, including `MakeGenericType`. I believe it is safe to use it in this context, but I need to verify it.